### PR TITLE
Request mutability

### PR DESCRIPTION
**Pull request by fafhrd91**
this is wip refactoring for request mutability.

1. Middleware trait needs some more changes
2. FromRequest is still old

@DoumanAsh 
@mitsuhiko

### CODE REVIEW

**Comment by fafhrd91** [comment]
main points

1. `RequestContext` - http request
2. `RequestState` - holds information about routing process
3. `Middleware::start()` accepts RequestContext and RequestState
4. `HttpRequest` does not mutate its data

missing:

FormRequest is still old impl, we need to switch to RequestContext and RequestState. There is
problem though, how to impl HttpRequest extractor.

**Comment by mitsuhiko** [comment]
Going to review today. Was busy the last few days :(

File: src/server/h1writer.rs [hunk 1]
```diff
@@ -166,16 +169,29 @@ impl<T: AsyncWrite, H: 'static> Writer for H1Writer<T, H> {
...
+                ResponseLength::Zero => {
+                    buffer.extend_from_slice(b"\r\ncontent-length: 0\r\n")
+                }
+                ResponseLength::Length(len) => {
+                    helpers::write_content_length(len, &mut buffer)
+                }
+                ResponseLength::Length64(len) => {
+                    let s = format!("{}", len);
+                    buffer.extend_from_slice(b"\r\ncontent-length: ");
+                    buffer.extend_from_slice(s.as_ref());
```
Comment (DoumanAsh) [inline src/server/h1writer.rs:185]: Does buffer implement `Write` trait so that we could use `write_fmt`?

**Review by DoumanAsh — COMMENTED** [review]
I did a quick look today, overall it looks good, I didn't leave much of comments as there seems to lots of unfinished stuff.
I especially like that in the end you moved away from separate `RequestState` param in middleware.
I think it would be the best to keep middleware as simple as possible.
I wonder will `response` and `finished` take it too instead of just `Request`? I do guess it is unfinished part too?

I'm also wondering whether are you going to remove `HttpRequest` or will we have separate `Request` struct? It doesn't really seems convenient to have various request types as it may confuse.

My current concern though is `RefCell` which potentially can lead to panic but from what I can see in code it seems that intention is that only one mutable borrow should present at the time.
Since `RefCell` implies runtime I'm wondering do you plan eventually get rid of it?

File: src/server/h1writer.rs (hunk 1 above)
Comment (fafhrd91) [reply src/server/h1writer.rs:185]: `write_fmt` is slow

**Comment by fafhrd91** [comment]
PR is ready

at the moment we have to use `RefCell` because of `Clone` for `HttpRequest`. async/await may simplify `Clone` requirements. we'll see.

i removed `RequestContext` it is not very ergonomic.

File: src/server/h1writer.rs (hunk 1 above)
Comment (jnicholls) [reply src/server/h1writer.rs:185]: But you are freely using `format!` here to stringify an integer?

Comment (fafhrd91) [reply src/server/h1writer.rs:185]: Missed that, too many changes. I will fix it

Comment (jnicholls) [reply src/server/h1writer.rs:185]: Yeah it's all good, there aren't many occasions the response size is that large (not enough per-second to make this string allocation a problem at least).

**Pull request merged**